Update version by changing a symlink

haslersn opened the pull request:
NOTE: This PR is exclusively alternative to #9.

Previously, mattermost was never updated after its initial installation. See #7.

Now we install mattermost in a separate folder per version. We check the if the configured version is installed and if not, install it (*). The newly installed version is then deployed by simply changing the symlink at `/opt/mattermost`.

(*) In https://docs.mattermost.com/administration/upgrade.html it's suggested to keep certain folders when updating, so we sync those from the current version to the newly installed version.

**Consider that, although it doesn't seem like it, this way of updating mattermost is less complex than the way in #9** (compare +116 -63 to +122 -61 changes).

Closes #7 
Closes #9

haslersn committed:
install: Create a `version` file

The created file contains the `mattermost` version. This enables the
possibility to fix #7 in the future.

haslersn committed:
Update version by changing a symlink

Previously, mattermost was never updated after its initial installation.
See #7.

Now we install mattermost in a separate folder per version. We check the
if the configured version is installed and if not, install it (*). The
newly installed version is then deployed by simply changing the symlink at
`/opt/mattermost`.

(*) In https://docs.mattermost.com/administration/upgrade.html it's suggested
to keep certain folders when updating, so we sync those from the current
version to the newly installed version.

haslersn committed:
configure: Remove duplications

Three tasks were previously unnecessarily duplicated. Also, setting
users and groups prior to the task 'Change Mattermost directory
permissions' is redundant.

SlothOfAnarchy commented on tasks/install.yml:
```diff
@@ -1,38 +1,24 @@
-  delegate_to: localhost
+- name: Check if the desired mattermost version exists
+  stat:
+    path: /opt/mattermost-versions/{{ mattermost_version }}
```
Won't this task fail if `/opt/mattermost-versions` doesn't exist, i.e. the first time this role is run on a server?

haslersn replied:
Shouldn't stat can handle folders that don't exist and can therefore be used to check whether a folder exists at all.

I will test it later, though.

haslersn replied:
I just tested it and it works like desired.

SlothOfAnarchy replied:
Okay, great. I just tested it using the `stat` command, which returned 1 in such a case.

SlothOfAnarchy commented on tasks/update.yml:
```diff
@@ -0,0 +1,31 @@
+    mattermost_archive_name: mattermost-{{ mattermost_version }}
+  when: mattermost_enterprise
+
+- name: 'Download binary from https://releases.mattermost.com'
```
I don't think this needs to be quoted.
Great idea to combine the tasks for both flavors!

haslersn replied:
This line wasn't changed, but you're right.

SlothOfAnarchy replied:
I know, I checked where the quotes were coming from

SlothOfAnarchy commented on tasks/update.yml:
```diff
@@ -0,0 +1,31 @@
+  unarchive:
+    src: "{{ global_cache_dir }}/{{ mattermost_archive_name }}-linux-amd64.tar.gz"
+    dest: /tmp
+  args:
```
Please remove the empty args key.

SlothOfAnarchy commented on tasks/update.yml:
```diff
@@ -0,0 +1,31 @@
+- name: Unpack Mattermost archive
+  unarchive:
+    src: "{{ global_cache_dir }}/{{ mattermost_archive_name }}-linux-amd64.tar.gz"
+    dest: /tmp
```
Why not unpack the archive directly to `/opt/mattermost-versions/{{ mattermost_version }}`?

haslersn replied:
Because `install.yml` assumes that every folder inside `/opt/mattermost-versions` contains a fully and successfully installed version.

SlothOfAnarchy commented on tasks/deploy.yml:
```diff
@@ -0,0 +1,33 @@
+- name: Set the symlink to the installed mattermost version
+  file:
+    src: /opt/mattermost-versions/{{ mattermost_version }}
+    dest: /tmp/mattermost
```
Is this step really necessary? Why not override the original symlink, shouldn't that also be atomic?
If you have a reason to keep it this way, please rename the symlink so it won't be confused with the extracted mattermost archive in `update.yml`.

haslersn replied:
No, changing a symlink involvis the system calls:
  * unlink
  * symlink

See: http://blog.moertel.com/posts/2005-08-22-how-to-change-symlinks-atomically.html

2. Ok.

SlothOfAnarchy reviewed (changes requested):
A bit complex to grasp at first, but really the smartest way to do it!

haslersn committed:
update: Remove unnecesary quotes and keys

haslersn committed:
main: Move mattermost_preserve_on_update to vars/main.yml

We don't see any need to configure the variable
`mattermost_preserve_on_update`, so we moved it to `vars/main.yml`

SlothOfAnarchy commented:
Please indicate that this role supports upgrading in `README.md`

haslersn committed:
README: Add section about updating

Updating mattermost via the role is possible since
8e5144380ecb4332d639d514c3579c56325a6a16. Now this is documented
in `README.md`.
